feat: notifications: report failed workspace builds

mtojek opened the pull request:
Fixes: https://github.com/coder/team-coconut/issues/40

This PR introduces support for email reports, including a report with failed workspace builds. A report generator runs periodically to see if users expect a new report. Once the report is successfully generated and spread, a record (or records) are logged in the database table.

In the multi-instance setup, report generators use an exclusive locking mechanism to ensure only one instance is running.

mtojek committed:
just the template

mtojek committed:
Bobby

mtojek committed:
golden

mtojek committed:
WIP

mtojek committed:
generator

mtojek committed:
lock

mtojek committed:
notif data

mtojek committed:
fix

mtojek committed:
fixmes

mtojek committed:
NoopEnqueuer

mtojek committed:
fixme helpers

mtojek committed:
Run generator

mtojek committed:
fix

mtojek committed:
TODO

mtojek committed:
naming

mtojek committed:
fix lint

mtojek committed:
TODO

mtojek committed:
fix dbmock

mtojek committed:
WIP

mtojek committed:
TODOs

mtojek committed:
WIP

mtojek committed:
fix

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
another WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
before input data

mtojek committed:
Last TODO

mtojek committed:
WIP

mtojek committed:
rebuild model

mtojek committed:
Merge branch 'main' into 40-failure-summary

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
Fixed

mtojek committed:
dbauthz

mtojek committed:
fix

mtojek committed:
fix

mtojek committed:
excluded

mtojek committed:
makegen

mtojek committed:
wIP

mtojek committed:
fix

mtojek committed:
fix

mtojek committed:
makegen

mtojek committed:
dbmem

mtojek committed:
makegen

mtojek committed:
last dbmem

mtojek committed:
fix

mtojek committed:
fix

mtojek committed:
fix

mtojek committed:
wip

mtojek committed:
fix

mtojek committed:
stub internal test

mtojek committed:
WIP

mtojek committed:
makegen

mtojek committed:
WIP

mtojek committed:
WIP: tests

mtojek committed:
cleanup

mtojek committed:
InitialState_NoBuilds_NoReport

mtojek committed:
first semi test

mtojek committed:
fix: created_by

mtojek committed:
fix: created_by tv

mtojek committed:
fix: build number

mtojek committed:
expanding

mtojek committed:
fix: snapshot

mtojek committed:
test done

mtojek committed:
more tests

mtojek committed:
WIP

mtojek committed:
tests done

mtojek committed:
cleanup

mtojek committed:
Merge branch 'main' into 40-failure-summary

dannykopping commented on coderd/database/lock.go:
```diff
@@ -10,6 +10,7 @@ const (
 	LockIDEnterpriseDeploymentSetup
 	LockIDDBRollup
 	LockIDDBPurge
+	LockIDReportGenerator
```
I think we should name this more specifically.

```suggestion
	LockIDNotificationsReportGenerator
```

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+VALUES ('34a20db2-e9cc-4a93-b0e4-8569699d7a00', 'Report: Workspace Builds Failed For Template', E'Workspace builds failed for template "{{.Labels.template_display_name}}"',
+        E'Hi {{.UserName}},
+
+Template **{{.Labels.template_display_name}}** has failed to build {{.Data.failed_builds}}/{{.Data.total_builds}} times over the last {{.Data.report_frequency}} and may be unstable.
```
Nit:
```suggestion
Template **{{.Labels.template_display_name}}** has failed to build {{.Data.failed_builds}}/{{.Data.total_builds}} times over the last {{.Data.report_frequency}}.
```

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+{{range $version := .Data.template_versions}}
+**{{$version.template_version_name}}** failed {{$version.failed_count}} time{{if gt $version.failed_count 1}}s{{end}}:
+{{range $build := $version.failed_builds}}
+* [{{$build.workspace_owner_username}} / {{$build.workspace_name}} / #{{$build.build_number}}]({{base_url}}/@{{$build.workspace_owner_username}}/{{$build.workspace_name}}/builds/{{$build.build_number}})
```
Nit: I think a table display might be nicer here.

dannykopping replied:
I'm not aware of any problems with tables in email; could you elaborate? Is it a technical issue, or more of a UX one?

mtojek replied:
UX, this is how I perceive tables used in emails (google some basic samples), especially if you don't know the size of the table

1. You need to scroll the screen:

<img width="500" alt="Screenshot 2024-09-13 at 15 51 24" src="https://github.com/user-attachments/assets/b3badfaa-7160-4b8d-b998-15b69a9c4b00">

2. Squized:

<img width="617" alt="Screenshot 2024-09-13 at 15 51 31" src="https://github.com/user-attachments/assets/054bbddb-2dc8-42af-8fd6-a157be4d1a86">

BTW What do you think is the best table for this use case? A few tables, one per template version? or a single big one? I'm trying to visualize it.

mafredri commented on coderd/database/queries/workspacebuilds.sql:
```diff
@@ -179,3 +179,62 @@ WHERE
+    templates AS t ON
+	w.template_id = t.id
+WHERE
+    wb.created_at > @since
```
```suggestion
    wb.created_at >= @since
```

Minor nit, "since" is inclusive, otherwise "after" works better.

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
+VALUES ('34a20db2-e9cc-4a93-b0e4-8569699d7a00', 'Report: Workspace Builds Failed For Template', E'Workspace builds failed for template "{{.Labels.template_display_name}}"',
```
Does this mean if I have 20 templates I could get up to 20 emails?

mafredri commented on coderd/database/queries/workspacebuilds.sql:
```diff
@@ -179,3 +179,62 @@ WHERE
+	wb.template_version_id = tv.id
+WHERE
+	w.template_id = $1
+	AND wb.created_at > @since
```
```suggestion
	AND wb.created_at >= @since
```

Same minor nit again.

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+    last_generated_at timestamp with time zone NOT NULL,
+
+    PRIMARY KEY (user_id, notification_template_id),
+    UNIQUE (user_id, notification_template_id)
```
I believe this is redundant because you're already using these fields in the PK.

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+    UNIQUE (user_id, notification_template_id)
+);
+
+COMMENT ON TABLE report_generator_logs IS 'Logs with generated reports for users.';
```
```suggestion
COMMENT ON TABLE report_generator_logs IS 'Log of generated reports for users.';
```

dannykopping commented on coderd/database/migrations/000250_email_reports.up.sql:
```diff
@@ -0,0 +1,32 @@
+        }
+    ]'::jsonb);
+
+CREATE TABLE report_generator_logs
```
```suggestion
CREATE TABLE notification_report_generator_logs
```

dannykopping commented on coderd/database/queries/notifications.sql:
```diff
@@ -174,3 +174,23 @@ SELECT *
 WHERE kind = @kind::notification_template_kind
 ORDER BY name ASC;
+
+-- name: GetReportGeneratorLogByUserAndTemplate :one
```
Nit: these functions should all be namespaced under "notifications" since it's a generic term.

```suggestion
-- name: GetNotificationReportGeneratorLogByUserAndTemplate :one
```

dannykopping replied:
My point here is that these reports are only used for notifications; "report" is such a generic term, and having these in the global scope along with all other queries will be confusing without being scoped to notifications.

mtojek replied:
Ok, will rename the package and relevant structures 👍

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+
+	"github.com/google/uuid"
+
+	"github.com/coder/quartz"
```
Dumb formatter/linter, imports are a mess 😔.

dannykopping commented on coderd/notifications/enqueuer.go:
```diff
@@ -52,9 +52,15 @@ func NewStoreEnqueuer(cfg codersdk.NotificationsConfig, store Store, helpers tem
 
+// Enqueue queues a notification message for later delivery, assumes no structured input data.
+func (s *StoreEnqueuer) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
+	// "nil" data will be omitted while building the JSON payload.
```
Can you please clarify this comment?
IMHO we should never omit keys from the payload, since webhook receivers may be expecting them to always be present.

mtojek replied:
I agree, but wouldn't that be a problem with backward compatibility of the payload format? I _really_ don't want to implement magic to support both of them.

dannykopping replied:
You don't need to do anything for BC-compat; you just bump the version number of the payload by a minor version and you're done. We're not receiving this payload, only sending it, so we never have to unless marshal it.

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	ctx = dbauthz.AsSystemRestricted(ctx)
+
+	// Start the ticker with the initial delay.
+	ticker := clk.NewTicker(delay)
```
```suggestion
	ticker := clk.NewTicker(delay)
	ticker.Stop()
```

Small suggestion to convey intent since we run a doTick with ticker.Reset soon afterwards anyway.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	// Start the ticker with the initial delay.
+	ticker := clk.NewTicker(delay)
+	doTick := func(start time.Time) {
+		defer ticker.Reset(delay)
```
I don't think this is needed? The ticker ticks with the given interval without needing to reset AFAIK.

mtojek replied:
Right, but if the routine takes a significant amount of time then we don't need an immediate start of next routine. Also, I don't think this is a new pattern in the project source. See `dbpurge.go`.

dannykopping replied:
https://pkg.go.dev/time#Ticker

> The ticker will adjust the time interval or drop ticks to make up for slow receivers.

I believe that will account for this situation, no?

mtojek replied:
Interesting, do you know how ticker would know how much time the routine spent doing some logic?

johnstcn replied:
I think it depends if we want the reports to run every interval, or if we want at least interval between executions of report.

https://go.dev/play/p/ZOQOKPIWD58

```
2009-11-10 23:00:01 +0000 UTC m=+1.000000001
1.5s <-- thing take 500ms
2009-11-10 23:00:02.5 +0000 UTC m=+2.500000001
2s   <-- thing take 1000ms
2009-11-10 23:00:04.5 +0000 UTC m=+4.500000001
2.5s <-- thing takes 1500ms
```

If you comment out the `tick.Reset()` then you will see:

```
2009-11-10 23:00:01 +0000 UTC m=+1.000000001
1.5s <- thing take 500ms
2009-11-10 23:00:02 +0000 UTC m=+2.000000001
1.5s <- thing takes 1000ms
2009-11-10 23:00:03 +0000 UTC m=+3.000000001
1.5s <- thing takes 1500ms
2009-11-10 23:00:04 +0000 UTC m=+4.000000001
2s   <-- thing takes 2000ms, ticker slows down
```

mtojek replied:
I'm going to leave it as is, but applied the suggestion from Mathias:

```go
	ticker := clk.NewTicker(delay)
	ticker.Stop()
```

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+			// Acquire a lock to ensure that only one instance of the generator is running at a time.
+			ok, err := tx.TryAcquireLock(ctx, database.LockIDReportGenerator)
+			if err != nil {
+				return err
```
```suggestion
				return xerrors.Errorf("failed to acquire report generator lock: %w", err)
```

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
```
This seems like the only possibility at this time, why do we need this handling?

mtojek replied:
In the original design there was an idea to make it customizable, but right now I can definitely simplify it. Thanks for spotting this!

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+			err = reportFailedWorkspaceBuilds(ctx, logger, db, enqueuer, clk)
+			if err != nil {
+				logger.Debug(ctx, "unable to report failed workspace builds")
+				return err
```
```suggestion
				return xerrors.Errorf("unable to report failed workspace builds: %w", err)
```

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	return nil
+}
+
+const failedWorkspaceBuildsReportFrequencyDays = 7
```
Without the "week" check, usage could be simplified by using a duration here.

mtojek replied:
Hm.. I'm not sure if I understood your comment 😅 Could you elaborate, please?

mafredri replied:
```suggestion
const failedWorkspaceBuildsReportFrequency = 7 * time.Hour * 24
```
This was my suggestion 😄

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+				return err
+			}
+			if !ok {
+				logger.Debug(ctx, "unable to acquire lock for generating periodic reports, skipping")
+				return nil
+			}
+
+			err = reportFailedWorkspaceBuilds(ctx, logger, db, enqueuer, clk)
+			if err != nil {
+				logger.Debug(ctx, "unable to report failed workspace builds")
```
Nit: you're referring to reports (plural) and report (singular) when referring to the same thing, please pick one to avoid confusing operators.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+		for {
+			select {
+			case <-ctx.Done():
+				return
```
Maybe add a log here to indicate that it's done?

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	if err != nil {
+		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
+	}
+	sort.Slice(statsRows, func(i, j int) bool {
```
Suggestion: Add `ORDER` to query so we can omit this sorting. It's preferable if queries always have stable output, unless it needs to be omitted for performance reasons.

mtojek replied:
Ok! Thanks for the explanation. I was not sure if stable output is a best practice.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+			case <-ctx.Done():
+				return
+			case tick := <-ticker.C:
+				ticker.Stop()
```
Not sure why you need this?

mtojek replied:
We discussed the patter in the other thread, so resolving this.

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+			return failedBuilds[i].TemplateVersionName < failedBuilds[j].TemplateVersionName
+		}
+		return failedBuilds[i].WorkspaceBuildNumber > failedBuilds[j].WorkspaceBuildNumber
+	})
```
Suggestion: Would this sort order make sense in the query and could be omitted here?

mtojek replied:
Not sure to be honest, I have to validate it. The idea is to present recent failures at the top.

dannykopping commented on coderd/notifications/enqueuer.go:
```diff
@@ -52,9 +52,15 @@ func NewStoreEnqueuer(cfg codersdk.NotificationsConfig, store Store, helpers tem
 // Enqueue queues a notification message for later delivery.
 // Messages will be dequeued by a notifier later and dispatched.
-func (s *StoreEnqueuer) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
+func (s *StoreEnqueuer) EnqueueData(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, data map[string]any, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
```
Nit: this will be clearer.

```suggestion
func (s *StoreEnqueuer) EnqueueWithData(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, data map[string]any, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
```

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+func (i *reportGenerator) Close() error {
+	i.cancel()
```
We may have to wait until the report is fully generated even after the context is canceled. I suggest adding a short-circuit to `reportFailedWorkspaceBuilds` to check in each loop iter if the context is canceled.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	if err != nil {
+		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
+	}
+	sort.Slice(statsRows, func(i, j int) bool {
```
Why not sort in the database?

mtojek replied:
Will do

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+		var failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow
+		reportData := map[string]any{}
+
+		if stats.FailedBuilds > 0 {
```
Can you perhaps invert this and `continue`?
Looks like we do a bunch of work we don't need to if there are no failed builds.

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+		}
+
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
+		builds := templateVersions[c-1]["failed_builds"].([]map[string]any)
```
Suggestion to make referencing easier on the eyes:

```
tv := templateVersions[c-1]
// ...
templateVersions[c-1] = tv
```

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+				UserID:                 templateAdmin.ID,
+				NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			})
+			if err != nil && !xerrors.Is(err, sql.ErrNoRows) { // sql.ErrNoRows: report not generated yet
```
Under what circumstances will this occur?

mtojek replied:
What do you mean? A generic error or `sql.ErrNoRows`?

Re `sql.ErrNoRows`: First generator run for a user. No log entries have been stored yet.

dannykopping replied:
> Re sql.ErrNoRows: First generator run for a user. No log entries have been stored yet.

Is that an error case?

mtojek replied:
Removed `logger.Error`.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+				continue
+			}
+
+			reportGeneratedNow[templateAdmin.ID] = true
```
Should this not only be set after checking if there are failed builds?

mtojek replied:
I think so. It indicates that some template admins will receive reports, and system should store the timestamp when it happened.

Did you have a specific use case in mind or is it just a general question?

dannykopping replied:
Well if `failedBuilds == 0` then it'lll still enter into the loop on line 181, no?

mtojek replied:
This seems to be a naming problem, `reportGeneratedNow` indicates that the report has been generated, but in fact it means that _the job processed the template admin and the template, and should check again in a week_ (it does not matter if the report was created or not).

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
+			templateVersions = append(templateVersions, map[string]any{
```
Is the reason we use `map[string]any` vs a well-known struct that we need the format for passing on to template? Otherwise I'd prefer the struct for this data, especially if its JSON encoded should be fine.

mtojek replied:
This needs to go through notifications plumbing that expects `map[string]any`.  `EnqueueData` interface is used in many places, so I'm afraid I can't hardcode the model structure.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+func buildDataForReportFailedWorkspaceBuilds(frequencyDays int, stats database.GetWorkspaceBuildStatsByTemplatesRow, failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow) map[string]any {
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
+	} else {
+		var plural string
+		if frequencyDays > 1 {
+			plural = "s"
+		}
+		frequencyLabel = fmt.Sprintf("%d day%s", frequencyDays, plural)
+	}
```
See `humanize.Time`.

mtojek replied:
I'm familiar with this concept/lib, but I believe it would be an overkill here?

dannykopping replied:
We already use the lib, I mean.

mtojek replied:
Resolving, as decided to stick with const "week".

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
```
```suggestion
		if c == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
```

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
```
Is this to prevent duplicates?

mtojek replied:
Not sure what you mean by _duplicates_.

`failedBuilds` are sorted by template versions ASC, so whenever there is a new template version, we append an entry to `templateVersions := []map[string]any{}`.

mafredri commented on coderd/notifications/testdata/rendered-templates/TemplateWorkspaceBuildsFailedReport-body.md.golden:
```diff
@@ -0,0 +1,17 @@
+Hi Bobby,
+
+Template **Bobby First Template** has failed to build 4/55 times over the last week and may be unstable.
```
_may be unstable_ love the optimism! 😄

mafredri reviewed (approved):
Mainly a logic review (not so much behavior), but nothing major stuck out. Small nits but otherwise LGTM!

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+			continue
+		}
+
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
```
Some comments here would be appreciated; it's difficult to understand what you're trying to do here.

mtojek replied:
I will drop detailed comments.

The idea is to review workspace builds, and link them with matching template versions.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	}
+
+	var templateAdmins []database.GetUsersRow
+	if len(userIDs) > 0 {
```
Nit: Invert and return early
The go idiom is that the most unintended code should be the happy path.

dannykopping commented on coderd/notifications/reports/generator_internal_test.go:
```diff
@@ -0,0 +1,286 @@
+		}
+
+		// Given: 6 days later (less than report frequency), and failed build
+		clk.Advance(6 * dayDuration).MustWait(context.Background())
```
Awesome, thanks for using quartz 👌 makes things much simpler.

dannykopping commented on coderd/notifications/types/payload.go:
```diff
@@ -17,4 +17,5 @@ type MessagePayload struct {
 
 	Actions []TemplateAction  `json:"actions"`
 	Labels  map[string]string `json:"labels"`
+	Data    map[string]any    `json:"data,omitempty"`
```
For interface stability.

```suggestion
	Data    map[string]any    `json:"data"`
```

mtojek replied:
Can you confirm that it will not affect backward compatibility with existing notifications?

dannykopping replied:
You're only adding an extra field, so I believe that will not cause any BC-breaks.

mafredri commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
```
Just realized, is this check sufficient? Template version names are only unique by template ID, so without checking the ID we might be conflating two versions?

mtojek replied:
It is a very good point! Fortunately, we're running `buildDataForReportFailedWorkspaceBuilds` per template, so we won't browse template versions belonging to a different template.

Does it answer your question?

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	delay = 15 * time.Minute
+)
+
+func NewReportGenerator(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) io.Closer {
```
I don't see any tests which exercise this, which makes me wonder how you've validated the locking?

mtojek replied:
You are correct, I haven't. Do you have a specific idea of what unit test would fulfill the gap? I could run two instances of Report Generator, but then what would we like to validate and how?

dannykopping replied:
Running multiple instances sounds good like a good approach.
You should aim to achieve a moderately high % of coverage since this is a crucial area. Achieving the coverage will require provoking the various corner-cases, like when a lock is already held when trying to run a report.

mtojek replied:
I see what you mean, but I have no clue how we can test locking without some _blackbox/timeout_ tests or artificially mocked database calls. I'm happy to pair-program on this with you.

IMHO the most crucial part of this implementation is making sure that consecutive runs (immediate runs) won't send reports. What is not tested using the current approach is exclusive locking, and this will be hard.

dannykopping replied:
> What is not tested using the current approach is exclusive locking, and this will be hard.

What makes this particularly hard (asking for real, not rhetorically)? To me it seems like we'd need to invoke `NewReportGenerator` in two separate goroutines and just ensure that we only get what we're expecting, which you've implemented in other tests.

I think it's quite risky to have functionality which should be guarding against pathological behaviour which is not covered by a test.

dannykopping reviewed (commented):
Looking good!

I think we need some test coverage around the locking since that's super critical to the success of this feature.

Left a few nits & suggestions, but nothing major except the fact that this could generate _n_ reports per _frequency_ where _n = number of templates_. I think that could become quite jarring for template admins to receive multiple reports a week, and makes me think we should be combining all failing templates into a single report. I don't see the value in having separate emails for each.

johnstcn commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+	}
+}
+
+func findTemplateAdmins(ctx context.Context, db database.Store, stats database.GetWorkspaceBuildStatsByTemplatesRow) ([]database.GetUsersRow, error) {
+	users, err := db.GetUsers(ctx, database.GetUsersParams{
+		RbacRole: []string{codersdk.RoleTemplateAdmin},
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("unable to fetch template admins: %w", err)
+	}
+
+	usersByIDs := map[uuid.UUID]database.GetUsersRow{}
+	var userIDs []uuid.UUID
+	for _, user := range users {
+		usersByIDs[user.ID] = user
+		userIDs = append(userIDs, user.ID)
+	}
+
+	var templateAdmins []database.GetUsersRow
+	if len(userIDs) > 0 {
+		orgIDsByMemberIDs, err := db.GetOrganizationIDsByMemberIDs(ctx, userIDs)
+		if err != nil {
+			return nil, xerrors.Errorf("unable to fetch organization IDs by member IDs: %w", err)
+		}
+
+		for _, entry := range orgIDsByMemberIDs {
+			if slices.Contains(entry.OrganizationIDs, stats.TemplateOrganizationID) {
+				templateAdmins = append(templateAdmins, usersByIDs[entry.UserID])
+			}
+		}
+	}
+	sort.Slice(templateAdmins, func(i, j int) bool {
+		return templateAdmins[i].Username < templateAdmins[j].Username
+	})
+
+	return templateAdmins, nil
+}
```
1) What this function actually does is a bit hard to read. 
My read of it is:
- For a given template:
  - Query all users with the site-wide "template admin" role.
  - Fetch all organizations of which the above users are a member
  - Return only those users who are a member of that same organization as the given template.
Might make sense to add a clarifying comment?

2) It seems that this could be simplified significantly by adding the capability to query by `org_id` to `GetUsers`. I don't think this is a blocker, as my recollection is that modifying this query is rather non-trivial. It might be good to add as a follow-up though.

mtojek replied:
AFAIR I _borrowed_ the logic from `templateACL` (and later `provisionedrserver.go`). I can definitely leave a comment describing what is going on here.

Re 2: I do not want to dive into org related stuff if it is not necessary :)

johnstcn commented on coderd/notifications/reports/generator.go:
```diff
@@ -0,0 +1,296 @@
+func buildDataForReportFailedWorkspaceBuilds(frequencyDays int, stats database.GetWorkspaceBuildStatsByTemplatesRow, failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow) map[string]any {
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
+	} else {
+		var plural string
+		if frequencyDays > 1 {
+			plural = "s"
+		}
+		frequencyLabel = fmt.Sprintf("%d day%s", frequencyDays, plural)
+	}
```
nit: convert to switch?

mtojek replied:
Actually removed it entirely

mtojek committed:
fix: LockIDNotificationsReportGenerator

mtojek commented:
@dannykopping 

> I think that could become quite jarring for template admins to receive multiple reports a week, and makes me think we should be combining all failing templates into a single report. I don't see the value in having separate emails for each.

I believe this is what we originally designed in the RFC document. Assuming that templates are healthy in general, we should not expect too many reports. Personally, I would stick with current approach, with email targeting specific templates, so as an operator I can ignore some of them, but I'm open to suggestions.

FYI Redesigning it to a single aggregated email report will take some time to modify the logic.

dannykopping commented:
> @dannykopping
> 
> > I think that could become quite jarring for template admins to receive multiple reports a week, and makes me think we should be combining all failing templates into a single report. I don't see the value in having separate emails for each.
> 
> I believe this is what we originally designed in the RFC document. Assuming that templates are healthy in general, we should not expect too many reports. Personally, I would stick with current approach, with email targeting specific templates, so as an operator I can ignore some of them, but I'm open to suggestions.
> 
> FYI Redesigning it to a single aggregated email report will take some time to modify the logic.

Happy to stick with it as it is now, but I'm just imagining a scenario where something common to all templates causes a failure at some point in the week and admins receive one email per template. I suspect that if we don't implement a unified report now we'll need to do it in the future, but I'm happy to wait until the request comes in if you'd rather land this.

mtojek committed:
Address Mathias' feedback

mtojek committed:
makegen

mtojek committed:
fmt

mtojek committed:
more

mtojek committed:
reports

mtojek committed:
payload

mtojek committed:
nil-omitted

mtojek committed:
failed-lock

mtojek committed:
xerrors

mtojek committed:
c=0

mtojek committed:
log-debug

mtojek committed:
enqueue with data

mtojek committed:
remove logger

mtojek committed:
makegen

mtojek committed:
sort

mtojek committed:
makegen

mtojek committed:
GetFailedWorkspaceBuildsByTemplateID

mtojek committed:
usersByIDs

mtojek committed:
makegen

mtojek committed:
fix

mtojek committed:
processedUsers

mtojek committed:
comment

mtojek committed:
WIP

mtojek committed:
notifications

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
notifications

mtojek committed:
fmt

mtojek committed:
rephrase

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
fix

mtojek committed:
refactored

mtojek committed:
WIP

mtojek committed:
refactored

mtojek committed:
adjust tests

mtojek committed:
Merge branch 'main' into 40-failure-summary

mtojek committed:
duplicate migration

mtojek committed:
dbauthz

mtojek committed:
fix

mtojek committed:
makegen

mtojek committed:
fix

mtojek committed:
fix: migration down

mtojek committed:
xerrors.Is

mtojek committed:
limit builds

mtojek commented:
Alright, I addressed most of the feedback. Important changes:
* I decided to stick with the concept of email per template. Template admins should not receive emails if there are no failed workspace builds.
* Don't need to remove old entries in the `notification_report_generator_log` as right now there will be one record for `notifications.TemplateWorkspaceBuildsFailedReport`. We will see more if there are more report type defined.
* Reports are sent to template admins at the same time. If somebody disables notifications, they will receive the report in the next week. Actually, it simplified the tracking logic.

dannykopping commented on coderd/notifications/reports/generator.go:
```diff
@@ -98,73 +96,87 @@ func (i *reportGenerator) Close() error {
-				continue
-			}
+	for _, stats := range templateStatsRows {
+		if xerrors.Is(ctx.Err(), context.Canceled) {
+			logger.Debug(ctx, "context is canceled, quitting")
+			break
+		}
```
This is the idiomatic way to express this.

```suggestion
		select {
			case <-ctx.Done():
				logger.Debug(ctx, "context is canceled, quitting", slog.Error(ctx.Err())
				break
			default:				
		}
```

dannykopping reviewed (approved):
LGTM

Approving to unblock, but I really feel we should try add some testing around `NewReportGenerator`. I'm happy to land this first and have a follow-up PR which just focuses on that, since it could get complicated.

Excited to see this in action!

mtojek committed:
context cancelled

mtojek committed:
Merge branch 'main' into 40-failure-summary

mtojek committed:
fix migration
